src/storage/garbage_collector.cpp

```diff
@@ -129,54 +243,281 @@ void GarbageCollector::ProcessDeferredActions() {
   }
 }
 
-void GarbageCollector::TruncateVersionChain(DataTable *const table, const TupleSlot slot,
-                                            const transaction::timestamp_t oldest) const {
+void GarbageCollector::ProcessTupleVersionChain(DataTable *const table, TupleSlot slot,
+                                                std::vector<transaction::timestamp_t> *const active_txns) {
+  // Read the Version Pointer of this tuple
   const TupleAccessStrategy &accessor = table->accessor_;
-  UndoRecord *const version_ptr = table->AtomicallyReadVersionPtr(slot, accessor);
-  // This is a legitimate case where we truncated the version chain but had to restart because the previous head
-  // was aborted.
-  if (version_ptr == nullptr) return;
-
-  // We need to special case the head of the version chain because contention with running transactions can happen
-  // here. Instead of a blind update we will need to CAS and prune the entire version chain if the head of the version
-  // chain can be GCed.
-  if (transaction::TransactionUtil::NewerThan(oldest, version_ptr->Timestamp().load())) {
-    if (!table->CompareAndSwapVersionPtr(slot, accessor, version_ptr, nullptr))
-      // Keep retrying while there are conflicts, since we only invoke truncate once per GC period for every
-      // version chain.
-      TruncateVersionChain(table, slot, oldest);
+  UndoRecord *version_chain_head;
+  version_chain_head = table->AtomicallyReadVersionPtr(slot, accessor);
+
+  // If no chain is passed, nothing is collected
+  if (version_chain_head == nullptr) {
     return;
   }
 
+  // Otherwise collect as much as possible
+  // Don't collect version_chain_head though
+  auto active_txns_iter = active_txns->begin();
```

Comment:
I believe this overshadows the changes made in #325?

We want those changes for performance issues on skewed workloads.